// @azure/communication-react/dist/dist-esm/react-composites/src/composites/CallComposite/components/MutedNotification.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { Icon, mergeStyles, Stack, Text, useTheme } from '@fluentui/react';
import React from 'react';
import { useLocale } from '../../localization';
/**
 * Notify the user that they're muted.
 */
export function MutedNotification(props) {
    const locale = useLocale();
    const theme = useTheme();
    if (!props.speakingWhileMuted) {
        return React.createElement(React.Fragment, null);
    }
    return (React.createElement(Stack, { horizontal: true, horizontalAlign: "center" },
        React.createElement(Stack, { horizontal: true, className: mergeStyles(stackStyle(theme)) },
            React.createElement(Icon, { iconName: "Muted", className: mergeStyles(iconStyle(theme)) }),
            React.createElement(Text, { className: mergeStyles(textStyle(theme)), "aria-live": 'polite' }, locale.strings.call.mutedMessage))));
}
const stackStyle = (theme) => {
    return {
        background: theme.palette.black,
        opacity: 0.8,
        gap: `1rem`,
        padding: `1rem`,
        borderRadius: theme.effects.roundedCorner4,
        width: 'fit-content'
    };
};
const iconStyle = (theme) => {
    return {
        color: theme.palette.white
    };
};
const textStyle = (theme) => {
    return {
        color: theme.palette.white,
        fontSize: `1rem`
    };
};
